orttraining/orttraining/core/graph/gradient_builder.h

```diff
@@ -77,6 +77,7 @@ DECLARE_GRADIENT_BUILDER(GetPadGradient)
 DECLARE_GRADIENT_BUILDER(GetIdentityGradient)
 DECLARE_GRADIENT_BUILDER(GetPythonOpGradient)
 DECLARE_GRADIENT_BUILDER(GetScatterNDGradient)
+DECLARE_GRADIENT_BUILDER(GetATenOpGradient)
 
 DECLARE_GRADIENT_BUILDER(GetExternalGradient)
```

Comment:
I must be missing something... 
If ATenOp has gradientbuilder defined in python, it should take "GetExternalGradient" path. 
Only gradient-undefined ATenOp should take GetATenOpGradient path. 

I am missing the piece of code that is handling this logic.

Reply:
The logic is at the beginning of GetGradientForOp in gradient_builder_registry.cc. Actually we will check external gradient for all node, the key for each node is domain::op_type (domain::op_type::aten_name::aten_overload_name for ATenOp), if the key is registered, external gradient will be used, otherwide, it will go to gradient builder by Op type.